src/lib.rs

```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+            .collect();
+
+        let future = futures::future::join_all(all)
+            .map(move |res_v| {
+                let mut result = SpotRequestResult::default();
                 for (name, setup, user, res) in res_v {
-                    setup_fns.insert(name, setup);
-                    usernames.insert(name, user);
+                    result.setup_fns.insert(name, &setup);
+                    result.usernames.insert(name, user);
                     for sir in res {
                         if let Some(sir) = sir.spot_instance_request_id {
                             trace!(log, "activated spot request"; "id" => &sir);
-                            id_to_name.insert(sir.clone(), name);
-                            spot_req_ids.push(sir);
+                            result.id_to_name.insert(sir.clone(), name.into());
+                            result.spot_req_ids.push(sir);
                         }
                     }
                 }
 
-                // 2. wait for instances to come up
-                let mut req = rusoto_ec2::DescribeSpotInstanceRequestsRequest::default();
-                req.spot_instance_request_ids = Some(spot_req_ids.clone());
-
-                debug!(log, "waiting for instances to spawn");
-                let wait_for_instances = futures::future::loop_fn(
-                    id_to_name,
-                    move |mut id_to_name| {
-                        trace!(log, "checking spot request status");
-
-                        ec2.describe_spot_instance_requests(&req)
-                            .then(move |res| {
-                                if let Err(e) = res {
-                                    let msg = format!("{}", e);
-                                    if msg.contains("The spot instance request ID")
-                                        && msg.contains("does not exist")
-                                    {
-                                        trace!(log, "spot instance requests not yet ready");
-                                        return Ok(futures::future::Loop::Continue(id_to_name));
-                                    } else {
-                                        return Err(e)
-                                            .context(format!("failed to describe spot instances"))
-                                            .map_err(Error::from);
-                                    }
-                                }
-                                let res = res.expect("Err checked above");
+                result
+            })
+            .map_err(Error::from);
+        Box::new(future)
+    }
 
-                                let any_pending = res.spot_instance_requests
+    fn wait_for_instances<'a>(
+        log: &'a slog::Logger,
+        ec2: &'a Box<rusoto_ec2::Ec2>,
+        spot_req_ids: Vec<String>,
+        id_to_name: HashMap<String, &'a str>,
+        max_wait: &'a Option<time::Duration>,
```

Comment:
I think you should just take an `Option<time::Duration>` here. Copying it seems fine.